### PR TITLE
initial version of checksum based freshness

### PR DESCRIPTION
**Pull request by Xaeroxe**
Implementation for https://github.com/rust-lang/cargo/issues/14136 and resolves https://github.com/rust-lang/cargo/issues/6529

This PR implements the use of checksums in cargo fingerprints as an alternative to using mtimes. This is most useful on systems with poor mtime implementations.

This has a dependency on https://github.com/rust-lang/rust/pull/126930. It's expected this will increase the time it takes to declare a build to be fresh. Still this loss in performance may be preferable to the issues the ecosystem has had with the use of mtimes for determining freshness.

### CODE REVIEW

File: Cargo.toml
Comment (tgross35) [inline Cargo.toml:207]: Does this get used anywhere? Only sha256 is mentioned in these commits (there probably isn't much reason to support md5 anyway)

Comment (Xaeroxe) [reply Cargo.toml:207]: I'm not happy about supporting md5 myself, but it is something that my rustc PR could feasibly emit. I'm happy to remove it from both PRs if we can build a consensus on that.

Comment (Xaeroxe) [reply Cargo.toml:207]: This deserves more explanation. Md5 is used by rustc for debug info files, so the support for it was already present in rustc.

Comment (tgross35) [reply Cargo.toml:207]: Makes sense, just doesn’t look like it is used anywhere in this diff?

Comment (bjorn3) [reply Cargo.toml:207]: It's imported in src/cargo/core/compiler/fingerprint/mod.rs, which github hides by default due to the size of the diff.

Comment (Xaeroxe) [reply Cargo.toml:207]: It gets used here [Line 380 of src/cargo/core/compiler/fingerprint/mod.rs](https://github.com/rust-lang/cargo/pull/14137/files#diff-33fc5e62ca2fa6b54fd70249825e4f28dcc490317cd3a89e3c07aa62754406ceR380)

Comment (arlosi) [reply Cargo.toml:207]: The only reason that md5 and sha-1 were included as options for the debuginfo checksum was compatibility with PDB (pre-sha256 support) & DWARF5 debug formats.

Since we're making something new here, I don't think there's any reason to support deprecated hashing algorithms.

Comment (tgross35) [reply Cargo.toml:207]: > It gets used here [Line 380 of src/cargo/core/compiler/fingerprint/mod.rs](https://github.com/rust-lang/cargo/pull/14137/files#diff-33fc5e62ca2fa6b54fd70249825e4f28dcc490317cd3a89e3c07aa62754406ceR380)

🤦‍♂️ didn't see the collapsed file when I ctrl+F'd for it

Comment (Xaeroxe) [reply Cargo.toml]: md5 and sha1 are no longer supported for this. I've also added an option to use xxhash. xxhash is not cryptographically secure, but that's not necessary for this use case. So I figured it was worth comparing the performance of it to sha256.

Comment (Xaeroxe) [reply Cargo.toml]: xxhash has been swapped out for blake3. It was decided during the rustc MCP that cryptographically secure hashes are necessary in order for this freshness algorithm to be correct.

**Comment by Xaeroxe** [comment]
Merge conflicts resolved.

File: src/cargo/core/features.rs
Comment (epage) [inline src/cargo/core/features.rs:760]: Unstable features should also be documented at https://doc.rust-lang.org/nightly/cargo/reference/unstable.html

See `src/doc/src/reference/unstable.md`

File: src/cargo/util/command_prelude.rs
Comment (epage) [inline src/cargo/util/command_prelude.rs:711]: I think there might be a misunderstanding which makes me wonder how the tests are working. `self.flag` is for reading CLI flags from clap but no `checksum-freshness` CLI flag was created.  Instead there is `checksum-freshness` unstable feature flag that should be accessible as `gctx.cli_unstable().checksum_freshness`

File: src/cargo/util/command_prelude.rs
Comment (epage) [inline src/cargo/util/command_prelude.rs:749]: If we're defining a feature flag for this, we don't need to require `-Zunstable-options` (which means we should also stop having the tests set it)

File: src/cargo/util/context/mod.rs
Comment (epage) [inline src/cargo/util/context/mod.rs:94]: This looks like dead code?

File: tests/testsuite/freshness_checksum.rs
Comment (epage) [inline tests/testsuite/freshness_checksum.rs:18]: At minimum, could you structure your PR so its
- A commit with these tests without `-Zchecksum-freshness`
- A commit with the checksum work that also updates the tests to pass `-Zchecksum-freshness`

A big benefit to this is it shows to reviewers / the community how this feature is comparing to what was being done before

(sometimes, I also break out "adding an unstable feature" into its own commit which is the flag + docs)

File: tests/testsuite/freshness_checksum.rs
Comment (epage) [inline tests/testsuite/freshness_checksum.rs:26]: nit: could you do
```rust
p.cargo("build -Zchecksum-freshness")
```

Identifying quickly in the test case what command is being run makes it a lot easier to read the test.  This buries it compared to most tests.

File: tests/testsuite/freshness_checksum.rs
Comment (epage) [inline tests/testsuite/freshness_checksum.rs:24]: Is there a reason we need `cargo build` instead of `cargo check`?  The latter helps keep test times down

File: tests/testsuite/freshness_checksum.rs
Comment (epage) [reply tests/testsuite/freshness_checksum.rs:18]: Note: I've not dug deep into the tests, waiting on this change

File: src/cargo/core/compiler/fingerprint/mod.rs
Comment (epage) [inline src/cargo/core/compiler/fingerprint/mod.rs:1523]: Why did you go with separate enums?  Are we not able to read the checksum feature flag?  Would it be better to have a `checksum: bool` in it?

File: src/cargo/core/compiler/fingerprint/mod.rs
Comment (epage) [inline src/cargo/core/compiler/fingerprint/mod.rs:2006]: nit: I'd make `path` the last parameter.  The first three are all related in determining what the `root` should be

Alternatively, you could just make this a `path_root` function that only takes the first three and the join is done in the caller

File: src/cargo/core/compiler/fingerprint/mod.rs
Comment (epage) [inline src/cargo/core/compiler/fingerprint/mod.rs:913]: This name doesn't tell the reader what this function is doing

Comment (epage) [inline src/cargo/core/compiler/fingerprint/mod.rs:913]: Not thrilled with using `Either`, I feel like its obscuring what this is doing.

Comment (epage) [reply src/cargo/core/compiler/fingerprint/mod.rs:913]: For both of these, the problems go away I think if we merge the `LocalFingerprint` variants

File: src/cargo/core/compiler/fingerprint/mod.rs
Comment (epage) [reply src/cargo/core/compiler/fingerprint/mod.rs:1523]: Ah, I assume this is tied to the serialization.  I wonder if we should have a slight decoupling as I feel like this makes things more complicated.  I hope having two types of fingerprints is short term and so we shouldn't try to over-generalize

File: Cargo.toml
Comment (ReillyBrogan) [reply Cargo.toml]: Blake3 appears to more mature over-all anyway as it has runtime CPU feature detection for SSE2/SSE4.1/AVX2/AVX-512. The most commonly used XXH library only supports compile-time configuration of these which makes it less suited to a tool like cargo that's compiled once and ran on a wide variety of hardware.

When I was benchmarking algorithms for a similar purpose (source fingerprinting in go-task) both Blake3 and XXH were fast enough that the bottleneck was almost always disk IO and the runtime differences were minimal.

File: src/cargo/core/compiler/fingerprint/mod.rs
Comment (Xaeroxe) [reply src/cargo/core/compiler/fingerprint/mod.rs:1523]: Can you help be better understand what such a decoupling might look like?

Comment (Xaeroxe) [reply src/cargo/core/compiler/fingerprint/mod.rs:1523]: Nevermind, I think I've got it.

File: tests/testsuite/freshness_checksum.rs
Comment (Xaeroxe) [reply tests/testsuite/freshness_checksum.rs:18]: So it's worth noting that `freshness_checksum.rs` derives very heavily from `freshness.rs`. So a version of `freshness_checksum.rs` without the freshness flag would just be a subset of `freshness.rs`. I'm not sure how this provides new information. There are two tests which are truly unique to `freshness_checksum.rs`, which are `same_size_different_content()` and `checksum_actually_uses_checksum()`.

One might debate the merit of duplicating the tests like that. If you really wanted to deduplicate the tests then this would likely require a special case be added to the test runner code.

File: src/cargo/core/features.rs
Comment (Xaeroxe) [reply src/cargo/core/features.rs:760]: Resolved in 6cf92e1f1677ebe030ddfabea8782e2f721378ce

File: src/cargo/util/command_prelude.rs
Comment (Xaeroxe) [reply src/cargo/util/command_prelude.rs]: Resolved in 6cf92e1f1677ebe030ddfabea8782e2f721378ce, this was just vestigial.

File: src/cargo/util/command_prelude.rs
Comment (Xaeroxe) [reply src/cargo/util/command_prelude.rs]: Resolved in 0240bf05d640eae07d6d74523e14eea303db983a

File: src/cargo/util/context/mod.rs
Comment (Xaeroxe) [reply src/cargo/util/context/mod.rs]: Resolved in 5afdc115d295f74c475a2252df40ff7d2351c3eb

File: src/cargo/core/compiler/fingerprint/mod.rs
Comment (Xaeroxe) [reply src/cargo/core/compiler/fingerprint/mod.rs]: Resolved in 5afdc115d295f74c475a2252df40ff7d2351c3eb

File: src/cargo/core/compiler/fingerprint/mod.rs
Comment (Xaeroxe) [reply src/cargo/core/compiler/fingerprint/mod.rs]: Function was removed as part of d422f64ca56a7b6dfab5c8e5d4bbb06cccbbe5b3

Comment (Xaeroxe) [reply src/cargo/core/compiler/fingerprint/mod.rs]: Function was removed as part of d422f64ca56a7b6dfab5c8e5d4bbb06cccbbe5b3

File: tests/testsuite/freshness_checksum.rs
Comment (Xaeroxe) [reply tests/testsuite/freshness_checksum.rs]: No, where it was possible to do so without breaking tests I downgraded these to `cargo check` in 3cc6814ee8ede4ab4f3defb581f37c0ae1a1623a

File: tests/testsuite/freshness_checksum.rs
Comment (Xaeroxe) [reply tests/testsuite/freshness_checksum.rs]: Resolved in 3cc6814ee8ede4ab4f3defb581f37c0ae1a1623a

File: src/cargo/core/compiler/fingerprint/mod.rs
Comment (Xaeroxe) [reply src/cargo/core/compiler/fingerprint/mod.rs]: Resolved in d422f64ca56a7b6dfab5c8e5d4bbb06cccbbe5b3

Comment (Xaeroxe) [reply src/cargo/core/compiler/fingerprint/mod.rs]: The original thinking was that I should maintain backwards compatibility with prior build caches, but then I remembered that cargo releases alongside rust, and rust doesn't provide backwards compatibility of build caches.

File: src/cargo/core/compiler/fingerprint/mod.rs
Comment (weihanglo) [reply src/cargo/core/compiler/fingerprint/mod.rs:2422]: As a side note: is it possible to combine a comment and the corresponding files into one line in rustc? It might make dep-info harder to read, though it is not meant to read by human, and combining them could potentially write less bytes. I dont know if it is really criticial in terms of performance.

```
src/lib.rs: src/someotherfile.rs # checksum:blake3=660e91cbd3 file_len:9
```

File: src/cargo/core/compiler/fingerprint/mod.rs
Comment (weihanglo) [inline src/cargo/core/compiler/fingerprint/mod.rs:2274]: (not a blocker) 
I feel like `files` and `checksum` should eventually merge into one field.

File: src/cargo/core/compiler/fingerprint/dirty_reason.rs
Comment (weihanglo) [inline src/cargo/core/compiler/fingerprint/dirty_reason.rs:271]: nit: arg capture whenver possible and making sense

```suggestion
                                "the file `{}` has changed (checksum didn't match, {stored_checksum} != {new_checksum})",
                                file.display(),
```

File: src/cargo/core/compiler/fingerprint/dirty_reason.rs
Comment (weihanglo) [inline src/cargo/core/compiler/fingerprint/dirty_reason.rs:190]: If we are not using `new` anywhere, should we just remove this field?

File: tests/testsuite/freshness_checksum.rs
Comment (weihanglo) [inline tests/testsuite/freshness_checksum.rs:1]: Oops. The test module might need to be rewritten. See #14039

We can sort this out later as follow-ups. Don't worry

Comment (weihanglo) [reply tests/testsuite/freshness_checksum.rs:1]: This file looks like a copy of `freshness.rs` with `-Zchecksum-freshness` and `masquerade_as_nightly_cargo` everywhere.

Could you point out which tests are new and worth a review? Guess they are

* `checksum_actually_uses_checksum`
* `same_size_different_content`
* `modifying_and_moving`

Also could you add a test that verifying `-Zchecksum-freshness` is gated behind nightly? (A test without `masquerade_as_nightly_cargo` is sufficient.

**Review by weihanglo — COMMENTED** [review]
Looks like we are almost there! Thank for yours efforts, epage and Xaeroxe. Thought it would be a tough review, but honestly it was a happy time :)

File: src/doc/src/reference/unstable.md
Comment (Xaeroxe) [reply src/doc/src/reference/unstable.md:511]: Addressed as part of ba5899ae0f2211184731b2a8a98313881cb85675

File: src/cargo/core/compiler/fingerprint/mod.rs
Comment (Xaeroxe) [reply src/cargo/core/compiler/fingerprint/mod.rs]: Addressed as part of ba5899ae0f2211184731b2a8a98313881cb85675

This wasn't a great use of this error type, so I added a new one.

File: src/cargo/core/compiler/fingerprint/mod.rs
Comment (Xaeroxe) [reply src/cargo/core/compiler/fingerprint/mod.rs]: Addressed as part of ba5899ae0f2211184731b2a8a98313881cb85675

File: src/cargo/core/compiler/fingerprint/mod.rs
Comment (Xaeroxe) [reply src/cargo/core/compiler/fingerprint/mod.rs]: Addressed in ba5899ae0f2211184731b2a8a98313881cb85675

File: src/cargo/core/compiler/fingerprint/mod.rs
Comment (Xaeroxe) [reply src/cargo/core/compiler/fingerprint/mod.rs]: Removed the macro as part of ba5899ae0f2211184731b2a8a98313881cb85675

File: src/cargo/core/compiler/fingerprint/mod.rs
Comment (Xaeroxe) [reply src/cargo/core/compiler/fingerprint/mod.rs]: Added as part of ba5899ae0f2211184731b2a8a98313881cb85675, thanks for the code suggestion!

File: src/cargo/core/compiler/fingerprint/mod.rs
Comment (Xaeroxe) [reply src/cargo/core/compiler/fingerprint/mod.rs]: Addressed in ba5899ae0f2211184731b2a8a98313881cb85675

File: src/cargo/core/compiler/fingerprint/dirty_reason.rs
Comment (Xaeroxe) [reply src/cargo/core/compiler/fingerprint/dirty_reason.rs]: Addressed in ba5899ae0f2211184731b2a8a98313881cb85675

File: src/cargo/core/compiler/fingerprint/dirty_reason.rs
Comment (Xaeroxe) [reply src/cargo/core/compiler/fingerprint/dirty_reason.rs]: Yeah, new is defined as precisely the opposite of `old`, so let's get rid of it. Done in ba5899ae0f2211184731b2a8a98313881cb85675

File: src/cargo/core/compiler/fingerprint/mod.rs
Comment (weihanglo) [reply src/cargo/core/compiler/fingerprint/mod.rs]: See <https://github.com/rust-lang/rust/pull/126930#discussion_r1783822618>, which shows that my suggestion is not even better.

File: src/cargo/core/compiler/fingerprint/mod.rs
Comment (Xaeroxe) [reply src/cargo/core/compiler/fingerprint/mod.rs]: They are now the same field

File: tests/testsuite/freshness_checksum.rs
Comment (weihanglo) [reply tests/testsuite/freshness_checksum.rs:1]: I've reworked on tests so they should be better to review commit-by-commit

* Updated to new test syntax.
* Split copying `freshness.rs` into its own commit.
* Split non-functional changes from the commit that enables `-Zfreshness-checksum`

Other than these, everything should be the same as before. I haven't added any nightly-gate tests.

Comment (weihanglo) [reply tests/testsuite/freshness_checksum.rs:1]: BTW this rework now passes all tests with <https://github.com/rust-lang/rust/pull/126930/commit/58c5ac43aea5c94d61e3a7a4e8646e714230574d> in <https://github.com/rust-lang/rust/pull/126930>. Feel free to continue.

Comment (Xaeroxe) [reply tests/testsuite/freshness_checksum.rs:1]: The nightly gate tests have been added. Thanks for your help with the other notes here!

**Comment by weihanglo** [comment]
@bors try

Let's see how it really goes.

**Comment by weihanglo** [comment]
All tests are passed. I will do another round of review today.

File: tests/testsuite/freshness_checksum.rs
Comment (weihanglo) [inline tests/testsuite/freshness_checksum.rs:2515]: This should be sufficient

```suggestion
#[cfg(all(target_arch = "x86_64", target_os = "windows", target_env = "msvc"))]
```

File: tests/testsuite/dep_info.rs
Comment (weihanglo) [inline tests/testsuite/dep_info.rs:1]: Could we have a test verifying that the Cargo-flavored dep-info under fingerprint folder is correctly encoded? Something like [`relative_depinfo_paths_ws`](https://github.com/rust-lang/cargo/blob/4ddd017694a0415a7c91598153e642e3c2b26321/tests/testsuite/dep_info.rs#L251) but simpler

**Review by weihanglo — COMMENTED** [review]
@Xanewok 

Would you be willing to rearrange commits and make them easier to review/track for future readers?

Some intermediate commits like a45d66345acad70e930f883a207cd1f9fb9b70b6 and  5b36d643b23172c9cbac05ccb47e217bf5b6f63b could be squashed, and some like what I've done for tests are helpful to understand they were copied from the other file.

If that sounds too much a burden, no worries. I can do a simple cleanup before this PR merges. Thank you :)

**Comment by Xaeroxe** [comment]
@weihanglo All comments addressed in 9ede54d000c09e1824080d02bfbee647009def96

Also the PR has been rebased to remove some of the rejected designs for this feature.

**Review by weihanglo — APPROVED** [review]
I just rearrange commits a bit, so we can remove unnecessary commit like 2dafe11009d3228612397c6fe87b02a5bef25b13.

Thank you for the hard work! I am going to merge this when the CI is green.

**Comment by jonhoo** [comment]
Is there a way we can tag this for mention in the release notes? May be worth giving a bit of celebration to :)

**Comment by weihanglo** [comment]
It has been listed in changelog of 1.83 already, under the nightly section! https://github.com/rust-lang/cargo/blob/master/CHANGELOG.md#cargo-183-2024-11-28

Though we haven't figured out the build script part (see <https://github.com/rust-lang/cargo/issues/14136>), maybe we could celebrate a bit later when we resolve that? :)

**Comment by weihanglo** [comment]
If anyone is interested in helping this go stabilized, please share your benchmark results for `-Zchecksum-freshness` on [this Zulip topic](https://rust-lang.zulipchat.com/#narrow/channel/334885-t-cargo.2Fbuild-integration/topic/Call.20for.20testing.3A.20use.20of.20checksums.20over.20mtime), or in #14722.

**Pull request merged**
